mongo/mongo.go

```diff
@@ -23,37 +23,69 @@
 package mongo
 
 import (
+	"context"
+	"fmt"
+	"strings"
+
+	jaeger "github.com/topfreegames/extensions/jaeger/mongo"
 	"github.com/topfreegames/extensions/mongo/interfaces"
-	mgo "gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
 )
 
 //Mongo holds the mongo database and connection
 // Mongo implements MongoDB interface
 type Mongo struct {
+	ctx     context.Context
 	session *mgo.Session
 	db      *mgo.Database
 }
 
 //NewMongo return Mongo instance with completed fields
 func NewMongo(session *mgo.Session, db *mgo.Database) *Mongo {
 	return &Mongo{
+		ctx:     nil,
 		session: session,
 		db:      db,
 	}
 }
 
+//WithContext creates a shallow copy that uses the given context
+func (m *Mongo) WithContext(ctx context.Context) interfaces.MongoDB {
+	return &Mongo{
+		ctx:     ctx,
+		session: m.session,
+		db:      m.db,
+	}
+}
+
 // Run executes run command on database
 func (m *Mongo) Run(cmd interface{}, result interface{}) error {
+	var err error
+
 	session := m.session.Copy()
 	defer session.Close()
-	return m.db.With(session).Run(cmd, result)
+
+	database := m.db.Name
+	args := formatArgs(cmd)
+
+	jaeger.Trace(m.ctx, database, database, "runCommand", args, func() error {
```

Comment:
Shouldn't this be `jaeger.Trace(m.ctx, database, collection...`?

Reply:
No, although it might be confusing. The second argument is actually the prefix for the method being called on the database, generally the full name of the collection, on the form `db.collection` . However, this specific method is not called on a collection, so the prefix is only `db`. 

Having said that, I will change the name of the parameter in question, so it is more appropriate.